woocommerce/woocommerce-gutenberg-products-block
# Add contexts for checkout

This is part of the extracting work from the reference pull #1780.

This pull does the following:

- restructures `@woocommerce/base-context` imports so it pulls from `index.js` in the `base/context` folder. That way tree-shaking will be used if possible. This also helps with organizing the folder structure a bit better for the new contexts.
- Fixes all the various imports through the code due to the above change.
- Adds `CheckoutContextProvider`, `useCheckoutContext` and related files.
- Adds `PaymentMethodDataProvider`, `usePaymentMethodDataContext` and related files.
- Adds `ShippingMethodDataProvider`, `useShippingMethodDataContext` and related files.
- Adds typedefs used for the various context interfaces.

For rationale and purpose of these contexts see the reference pull description (especially the section beginning with _General Concepts_ and _Details_. Besides that, note the following.

- These contexts are not fully implemented yet. Subsequent pulls will start wiring things up to various aspects of checkout.
- The new contexts currently only apply to the checkout block.
- `ShippingMethodDataProvider` and context still has some work to do once shipping method hooks are done (need error exposure etc).
- There are still some inline @todos in the new code to help with tracking work that needs done (which are also exposed in #1780).

I haven't added any tests yet but that will be an important step in subsequent pulls once the contexts are _implemented_ more fully.  With the stage we're at, I want to unblock other work before spending time in pulls like this on tests.

## To Test

The important thing for this pull is to verify the following:

- Checkout block loads on frontend and editor without errors.
- Cart block loads on frontend and editor without errors and works as expected.

## Next Steps

- After merge and once subsequent pulls being extracted from #1780 are done. I'll create issues for helping track the remaining work needed.
